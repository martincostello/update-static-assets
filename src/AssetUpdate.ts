// Copyright (c) Martin Costello, 2022. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { CdnPackage } from './CdnPackage';
import { CdnProvider } from './CdnProvider';

export interface AssetUpdate extends CdnPackage {
  cdn: CdnProvider;
  pullRequestNumber: number;
  pullRequestUrl: string;
  supersedes: number[];
}
