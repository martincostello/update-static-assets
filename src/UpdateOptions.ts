// Copyright (c) Martin Costello, 2022. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { IgnoreAsset } from './IgnoreAsset';

export interface UpdateOptions {
  accessToken: string;
  apiUrl?: string;
  branchPrefix: string;
  commitMessage: string;
  dryRun: boolean;
  fileExtensions: string[];
  ignore: IgnoreAsset[];
  labels: string;
  repo?: string;
  repoPath: string;
  runId?: string;
  serverUrl?: string;
  userEmail: string;
  userName: string;
}
