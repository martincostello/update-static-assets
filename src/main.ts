// Copyright (c) Martin Costello, 2022. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import * as fs from 'fs';
import * as path from 'path';

import { StaticAssetUpdater } from './StaticAssetUpdater';
import { UpdateOptions } from './UpdateOptions';

export async function run(): Promise<void> {
  try {
    let repoPath = core.getInput('repo-path', { required: false }) ?? '.';
    repoPath = path.normalize(repoPath);

    const options: UpdateOptions = {
      accessToken: core.getInput('repo-token', { required: true }),
      apiUrl: process.env.GITHUB_API_URL ?? 'https://api.github.com',
      branchPrefix: core.getInput('branch-name-prefix', { required: false }),
      commitMessage: core.getInput('commit-message', { required: false }),
      dryRun: core.getInput('dry-run', { required: false }) === 'true',
      fileExtensions: [],
      ignore: [],
      labels: core.getInput('labels', { required: false }) ?? '',
      repo: process.env.GITHUB_REPOSITORY,
      repoPath,
      runId: process.env.GITHUB_RUN_ID,
      serverUrl: process.env.GITHUB_SERVER_URL ?? 'https://github.com',
      userEmail: core.getInput('user-email', { required: false }),
      userName: core.getInput('user-name', { required: false }),
    };

    const extensions =
      core.getInput('file-extensions', { required: false }) ??
      'cshtml,html,razor';

    options.fileExtensions = extensions.split(',');

    let configFile = core.getInput('configuration-file', { required: false });

    if (!configFile) {
      configFile = path.join(options.repoPath, '.update-static-assets.json');
    }

    if (configFile) {
      configFile = path.normalize(configFile);
    }

    if (fs.existsSync(configFile)) {
      const config = JSON.parse(fs.readFileSync(configFile, 'utf8'));

      if (config.ignore) {
        options.ignore = config.ignore;
      }
    }

    const updater = new StaticAssetUpdater(options);
    const result = await updater.tryUpdateAssets();

    core.setOutput('assets-updated', result.updates.length > 0);
  } catch (error: any) {
    core.error('Failed to check for updates to static assets.');
    core.error(error);
    if (error instanceof Error) {
      core.setFailed(error.message);
    }
  }
}

if (require.main === module) {
  run();
}
