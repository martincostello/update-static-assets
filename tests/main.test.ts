// Copyright (c) Martin Costello, 2022. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { v4 as uuid } from 'uuid';

import * as core from '@actions/core';
import * as exec from '@actions/exec';
import * as io from '@actions/io';

const github = require('@actions/github');

import { run } from '../src/main';

import {
  afterAll,
  beforeAll,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';

const tempDir = path.join(os.tmpdir(), 'update-static-assets-temp');

describe('update-static-assets tests', () => {
  beforeAll(async () => {
    if (!fs.existsSync(tempDir)) {
      await io.mkdirP(tempDir);
    }
  });

  afterAll(async () => {
    try {
      await io.rmRF(tempDir);
    } catch {
      console.log('Failed to remove test directories');
    }
  }, 5000);

  describe('Updates static assets for cdnjs', () => {
    const repoPath = path.join(tempDir, uuid());
    const indexHtml = path.join(repoPath, 'index.html');

    const testFiles = [
      {
        path: indexHtml,
        data: `
        <html lang="en-gb">
          <head>
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
          </head>
        </html>`,
      },
    ];

    beforeAll(async () => {
      setupMocks();
      await runAction(repoPath, testFiles);
    }, 30000);

    afterAll(async () => {
      try {
        await io.rmRF(repoPath);
      } catch {
        console.log('Failed to remove test repository');
      }
    }, 5000);

    test('Does not log any errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
    });

    test('Does not fail', () => {
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('Updates the font-awesome tag', () => {
      const html = fs.readFileSync(indexHtml, { encoding: 'utf8' });
      expect(html).not.toContain(
        'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css'
      );
      expect(html).not.toContain(
        'sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg=='
      );
    });
  });

  describe('Updates static assets for jsDelivr', () => {
    const repoPath = path.join(tempDir, uuid());
    const indexHtml = path.join(repoPath, 'index.html');

    const testFiles = [
      {
        path: indexHtml,
        data: `
        <html lang="en-gb">
          <head>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
          </head>
          <body>
            <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
          </body>
        </html>`,
      },
    ];

    beforeAll(async () => {
      setupMocks();
      await runAction(repoPath, testFiles);
    }, 30000);

    afterAll(async () => {
      try {
        await io.rmRF(repoPath);
      } catch {
        console.log('Failed to remove test repository');
      }
    }, 5000);

    test('Does not log any errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
    });

    test('Does not fail', () => {
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('Updates the bootstrap tags', () => {
      const html = fs.readFileSync(indexHtml, { encoding: 'utf8' });
      expect(html).not.toContain(
        'https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css'
      );
      expect(html).not.toContain(
        'sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3'
      );
      expect(html).not.toContain(
        'https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js'
      );
      expect(html).not.toContain(
        'sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p'
      );
    });
  });

  describe('Updates the static assets for multiple assets', () => {
    const repoPath = path.join(tempDir, uuid());
    const indexHtml = path.join(repoPath, 'index.html');

    const testFiles = [
      {
        path: indexHtml,
        data: `
        <html lang="en-gb">
          <head>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
          </head>
          <body>
            <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
          </body>
        </html>`,
      },
    ];

    beforeAll(async () => {
      setupMocks();
      await runAction(repoPath, testFiles);
    }, 30000);

    afterAll(async () => {
      try {
        await io.rmRF(repoPath);
      } catch {
        console.log('Failed to remove test repository');
      }
    }, 5000);

    test('Does not log any errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
    });

    test('Does not fail', () => {
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('Updates bootstrap', async () => {
      const stdout = await git(repoPath, 'branch');
      const matches = stdout.match(
        /(update-static-assets\/bootstrap\/[0-9\.]+)/
      );

      expect(matches).not.toBeNull();
      expect(matches!.length).toBeGreaterThanOrEqual(1);

      await git(repoPath, 'checkout', matches![0]);

      const html = fs.readFileSync(indexHtml, { encoding: 'utf8' });

      expect(html).not.toContain(
        'https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css'
      );
      expect(html).not.toContain(
        'sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3'
      );
      expect(html).not.toContain(
        'https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js'
      );
      expect(html).not.toContain(
        'sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p'
      );
    });

    test('Updates font-awesome', async () => {
      const stdout = await git(repoPath, 'branch');
      const matches = stdout.match(
        /(update-static-assets\/font-awesome\/[0-9\.]+)/
      );

      expect(matches).not.toBeNull();
      expect(matches!.length).toBeGreaterThanOrEqual(1);

      await git(repoPath, 'checkout', matches![0]);

      const html = fs.readFileSync(indexHtml, { encoding: 'utf8' });

      expect(html).not.toContain(
        'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css'
      );
      expect(html).not.toContain(
        'sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg=='
      );
    });
  });
});

describe('Does not update static assets that are ignored', () => {
  const repoPath = path.join(tempDir, uuid());
  const indexHtml = path.join(repoPath, 'index.html');

  const testFiles = [
    {
      path: indexHtml,
      data: `
      <html lang="en-gb">
        <head>
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
        </head>
      </html>`,
    },
    {
      path: path.join(repoPath, '.update-static-assets.json'),
      data: JSON.stringify({
        ignore: [
          {
            cdn: 'cdnjs',
            name: 'font-awesome',
            version: '.*',
          },
        ],
      }),
    },
  ];

  beforeAll(async () => {
    setupMocks();
    await runAction(repoPath, testFiles);
  }, 30000);

  afterAll(async () => {
    try {
      await io.rmRF(repoPath);
    } catch {
      console.log('Failed to remove test repository');
    }
  }, 5000);

  test('Does not log any errors', () => {
    expect(core.error).toHaveBeenCalledTimes(0);
  });

  test('Does not fail', () => {
    expect(core.setFailed).toHaveBeenCalledTimes(0);
  });

  test('Does not update the font-awesome tag', () => {
    const html = fs.readFileSync(indexHtml, { encoding: 'utf8' });
    expect(html).toContain(
      'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css'
    );
    expect(html).toContain(
      'sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg=='
    );
  });
});

function setupMocks(): void {
  jest.spyOn(process.stdout, 'write').mockImplementation(() => true);
  jest.spyOn(core, 'error').mockImplementation(() => {});
  jest.spyOn(core, 'setFailed').mockImplementation(() => {});
}

async function git(repoPath: string, ...args: string[]): Promise<string> {
  const options = {
    cwd: repoPath,
    ignoreReturnCode: true,
  };
  const result = await exec.getExecOutput('git', args, options);
  return result.stdout;
}

async function createTestGitRepo(
  repoPath: string,
  testFiles: { path: string; data: string }[]
): Promise<void> {
  if (!fs.existsSync(repoPath)) {
    await io.mkdirP(repoPath);
  }

  for (const { path, data } of testFiles) {
    fs.writeFileSync(path, data);
  }

  await git(repoPath, 'init');
  await git(repoPath, 'config', 'core.safecrlf', 'false');
  await git(repoPath, 'config', 'user.email', 'test@test.local');
  await git(repoPath, 'config', 'user.name', 'test');
  await git(repoPath, 'add', '.');
  await git(repoPath, 'commit', '-m', 'Initial commit');
}

async function runAction(
  repoPath: string,
  testFiles: { path: string; data: string }[]
): Promise<void> {
  const inputs = {
    'GITHUB_API_URL': 'https://github.local/api/v3',
    'GITHUB_REPOSITORY': '',
    'GITHUB_SERVER_URL': 'https://github.local',
    'INPUT_FILE-EXTENSIONS': 'cshtml,html,razor',
    'INPUT_LABELS': 'foo,bar',
    'INPUT_REPO-PATH': repoPath,
    'INPUT_REPO-TOKEN': 'my-token',
    'INPUT_USER-EMAIL': 'github-actions[bot]@users.noreply.github.com',
    'INPUT_USER-NAME': 'github-actions[bot]',
  };

  for (const key in inputs) {
    process.env[key] = inputs[key as keyof typeof inputs];
  }

  await createTestGitRepo(repoPath, testFiles);

  github.getOctokit = jest.fn().mockReturnValue({
    rest: {
      issues: {
        addLabels: () => Promise.resolve({}),
      },
      pulls: {
        create: () =>
          Promise.resolve({
            data: {
              number: '42',
              html_url:
                'https://github.local/martincostello/update-static-assets/pull/42',
            },
          }),
      },
    },
  });

  await run();
}
