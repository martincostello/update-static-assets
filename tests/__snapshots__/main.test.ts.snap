// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update-static-assets for assets that are ignored does not update excluded assets 1`] = `
"
          <html lang="en-gb">
            <head>
              <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
            </head>
          </html>"
`;

exports[`update-static-assets for assets that are ignored generates the correct commit message 1`] = `
[
  "Initial commit",
]
`;

exports[`update-static-assets for cdnjs generates the correct commit message 1`] = `
[
  "Update font-awesome",
  "Updates font-awesome to version 6.4.2.",
  "---",
  "updated-dependencies:",
  "- dependency-name: font-awesome",
  "  dependency-type: direct:production",
  "  update-type: version-update:semver-minor",
  "...",
  "Initial commit",
]
`;

exports[`update-static-assets for cdnjs generates the correct diff 1`] = `
"diff --git a/index.html b/index.html
index c435ad4..d7429e2 100644
--- a/index.html
+++ b/index.html
@@ -1,6 +1,6 @@
 
           <html lang="en-gb">
             <head>
-              <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
+              <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/all.min.css" integrity="sha512-z3gLpd7yknf1YoNbCzqRKc4qyor8gaKU1qmn+CShxbuBusANI9QpRohGBreCFkKxLhei6S9CQXFEbbKuqLg0DA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
             </head>
           </html>
\\ No newline at end of file"
`;

exports[`update-static-assets for cdnjs updates font-awesome 1`] = `
"
          <html lang="en-gb">
            <head>
              <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/all.min.css" integrity="sha512-z3gLpd7yknf1YoNbCzqRKc4qyor8gaKU1qmn+CShxbuBusANI9QpRohGBreCFkKxLhei6S9CQXFEbbKuqLg0DA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
            </head>
          </html>"
`;

exports[`update-static-assets for jsDelivr generates the correct commit message 1`] = `
[
  "Update bootstrap",
  "Updates bootstrap to version 5.3.1.",
  "---",
  "updated-dependencies:",
  "- dependency-name: bootstrap",
  "  dependency-type: direct:production",
  "  update-type: version-update:semver-minor",
  "...",
  "Initial commit",
]
`;

exports[`update-static-assets for jsDelivr generates the correct diff 1`] = `
"diff --git a/index.html b/index.html
index 498e7ea..1a51386 100644
--- a/index.html
+++ b/index.html
@@ -1,9 +1,9 @@
 
           <html lang="en-gb">
             <head>
-              <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
+              <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/css/bootstrap.min.css" integrity="sha256-2TnSHycBDAm2wpZmgdi0z81kykGPJAkiUY+Wf97RbvY=" crossorigin="anonymous">
             </head>
             <body>
-              <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
+              <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/js/bootstrap.bundle.min.js" integrity="sha256-0upsHgyryiDRjpJLJaHNAYfDi6fDP2CrBuGwQCubzbU=" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
             </body>
           </html>
\\ No newline at end of file"
`;

exports[`update-static-assets for jsDelivr updates bootstrap 1`] = `
"
          <html lang="en-gb">
            <head>
              <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/css/bootstrap.min.css" integrity="sha256-2TnSHycBDAm2wpZmgdi0z81kykGPJAkiUY+Wf97RbvY=" crossorigin="anonymous">
            </head>
            <body>
              <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/js/bootstrap.bundle.min.js" integrity="sha256-0upsHgyryiDRjpJLJaHNAYfDi6fDP2CrBuGwQCubzbU=" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
            </body>
          </html>"
`;

exports[`update-static-assets for multiple assets updating bootstrap generates the correct commit message 1`] = `
[
  "Update bootstrap",
  "Updates bootstrap to version 5.3.1.",
  "---",
  "updated-dependencies:",
  "- dependency-name: bootstrap",
  "  dependency-type: direct:production",
  "  update-type: version-update:semver-minor",
  "...",
  "Initial commit",
]
`;

exports[`update-static-assets for multiple assets updating bootstrap generates the correct diff 1`] = `
"diff --git a/index.html b/index.html
index 9a61475..55520a1 100644
--- a/index.html
+++ b/index.html
@@ -1,10 +1,10 @@
 
         <html lang="en-gb">
           <head>
-            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
+            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/css/bootstrap.min.css" integrity="sha256-2TnSHycBDAm2wpZmgdi0z81kykGPJAkiUY+Wf97RbvY=" crossorigin="anonymous">
             <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
           </head>
           <body>
-            <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
+            <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/js/bootstrap.bundle.min.js" integrity="sha256-0upsHgyryiDRjpJLJaHNAYfDi6fDP2CrBuGwQCubzbU=" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
           </body>
         </html>
\\ No newline at end of file"
`;

exports[`update-static-assets for multiple assets updating bootstrap updates the HTML 1`] = `
"
        <html lang="en-gb">
          <head>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/css/bootstrap.min.css" integrity="sha256-2TnSHycBDAm2wpZmgdi0z81kykGPJAkiUY+Wf97RbvY=" crossorigin="anonymous">
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
          </head>
          <body>
            <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/js/bootstrap.bundle.min.js" integrity="sha256-0upsHgyryiDRjpJLJaHNAYfDi6fDP2CrBuGwQCubzbU=" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
          </body>
        </html>"
`;

exports[`update-static-assets for multiple assets updating font-awesome generates the correct commit message 1`] = `
[
  "Update font-awesome",
  "Updates font-awesome to version 6.4.2.",
  "---",
  "updated-dependencies:",
  "- dependency-name: font-awesome",
  "  dependency-type: direct:production",
  "  update-type: version-update:semver-minor",
  "...",
  "Initial commit",
]
`;

exports[`update-static-assets for multiple assets updating font-awesome generates the correct diff 1`] = `
"diff --git a/index.html b/index.html
index 9a61475..ab47560 100644
--- a/index.html
+++ b/index.html
@@ -2,7 +2,7 @@
         <html lang="en-gb">
           <head>
             <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
-            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0/css/all.min.css" integrity="sha512-9usAa10IRO0HhonpyAIVpjrylPvoDwiPUiKdWk5t3PyolY1cOd4DSE0Ga+ri4AuTroPR5aQvXU9xC6qOPnzFeg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
+            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/all.min.css" integrity="sha512-z3gLpd7yknf1YoNbCzqRKc4qyor8gaKU1qmn+CShxbuBusANI9QpRohGBreCFkKxLhei6S9CQXFEbbKuqLg0DA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
           </head>
           <body>
             <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous" referrerpolicy="no-referrer"></script>"
`;

exports[`update-static-assets for multiple assets updating font-awesome updates the HTML 1`] = `
"
        <html lang="en-gb">
          <head>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/all.min.css" integrity="sha512-z3gLpd7yknf1YoNbCzqRKc4qyor8gaKU1qmn+CShxbuBusANI9QpRohGBreCFkKxLhei6S9CQXFEbbKuqLg0DA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
          </head>
          <body>
            <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
          </body>
        </html>"
`;
